circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (C) 2016-2019 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.s3s3copier.aws;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.retry.annotation.Retryable;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.CopyObjectRequest;
+import com.amazonaws.services.s3.transfer.Copy;
+import com.amazonaws.services.s3.transfer.TransferManager;
+import com.amazonaws.services.s3.transfer.internal.TransferStateChangeListener;
+
+public class RetryableTransferManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryableTransferManager.class);
+  private static final int BACKOFF_DELAY_MS = 500;
+  private TransferManager transferManager;
+  private AmazonS3 srcClient;
+  private int retryAttempt = 0;
+
+  public RetryableTransferManager(TransferManager transferManager, AmazonS3 srcClient) {
+    this.transferManager = transferManager;
+    this.srcClient = srcClient;
+  }
+
+  @Retryable(value = { AmazonS3Exception.class },
+      backoff = @Backoff(delay = BACKOFF_DELAY_MS, multiplier = 2))
+  public Copy copy(CopyObjectRequest copyObjectRequest, TransferStateChangeListener stateChangeListener) {
+    retryAttempt++;
+    LOG.info("copying attempt {}/3", retryAttempt);
```

Comment:
Where does this 3 come from, is that a spring default?

Reply:
Yep Spring defaults to 3 attempts. I've also noticed that if there are multiple objects being copied, the `retryAttempt` will continue to increment. Adding a test for this now so that the logs are correct.

Reply:
ok, let's make the nr of attempts configurable. I guess we can't use the spring property as that would be the same for all RetryTemplates (we'd have a clash if we ever add another in the future??).

Reply:
Do you mean make it configurable for the user? If so I will have to refactor as annotation values must be constants and so aren't configurable.

And yes, with this setup, the spring property would apply across all retries. It may be possible to maintain multiple `RetryTemplate`s to avoid having clashes.

Reply:
Also, it would be nice to declare '3' as a constant (perhaps, MAX_RETRIES).

Reply:
@abhimanyugupta07 I've gone with `MAX_ATTEMPTS` as it includes the first failure :)

Reply:
And use that constant on line 49 in the log statment as well, please.

Reply:
ah yes, good spot, thanks!

Reply:
this doesn't work: ``` maxAttempts="${s3toS3copier-retry-maxattempts}"``` with that value coming from the yml? Might have to use: `maxAttemptsExpression`

Reply:
It doesn't look like it unfortunately, Java annotations must be constants. Have I understood you correctly? 

As a workaround, I could hardcode the default number of attempts and then break from the retry early if the user specifies a lower number of attempts?

Reply:
This has now been refactored so that the `maxAttempts` property is configurable